pkg/compact/downsample.go

```diff
@@ -0,0 +1,229 @@
+package compact
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"path/filepath"
+	"strings"
+
+	"github.com/improbable-eng/thanos/pkg/block"
+
+	"github.com/go-kit/kit/log"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/index"
+	"github.com/prometheus/tsdb/labels"
+)
+
+// Downsample downsamples the given block. It writes a new block into dir and returns its ID.
+func Downsample(ctx context.Context, b *tsdb.Block, dir string, window int64) (id ulid.ULID, err error) {
+	h, err := tsdb.NewHead(nil, nil, tsdb.NopWAL(), math.MaxInt64)
+	if err != nil {
+		return id, err
+	}
+	defer h.Close()
+
+	q, err := tsdb.NewBlockQuerier(b, b.Meta().MinTime, b.Meta().MaxTime)
+	if err != nil {
+		return id, errors.Wrap(err, "create block querier")
+	}
+	defer q.Close()
+
+	set, err := q.Select(labels.NewEqualMatcher(index.AllPostingsKey()))
+	if err != nil {
+		return id, errors.Wrap(err, "select all series")
+	}
+	for set.Next() {
+		s := set.At()
+
+		app := h.Appender()
+
+		if err := downsampleAggr(app, s.Labels(), s.Iterator(), window); err != nil {
+			return id, errors.Wrap(err, "downsample average")
+		}
+		if err := app.Commit(); err != nil {
+			return id, errors.Wrap(err, "commit")
+		}
+	}
+	if set.Err() != nil {
+		return id, errors.Wrap(set.Err(), "iterate series set")
+	}
+	comp, err := tsdb.NewLeveledCompactor(nil, log.NewNopLogger(), []int64{b.Meta().MaxTime - b.Meta().MinTime}, nil)
+	if err != nil {
+		return id, errors.Wrap(err, "create compactor")
+	}
+	id, err = comp.Write(dir, h, b.Meta().MinTime, b.Meta().MaxTime)
+	if err != nil {
+		return id, errors.Wrap(err, "compact head")
+	}
+	bdir := filepath.Join(dir, id.String())
+
+	origMeta, err := block.ReadMetaFile(b.Dir())
+	if err != nil {
+		return id, errors.Wrap(err, "read block meta")
+	}
+
+	meta, err := block.ReadMetaFile(bdir)
+	if err != nil {
+		return id, errors.Wrap(err, "read block meta")
+	}
+	meta.Thanos.Labels = origMeta.Thanos.Labels
+	meta.Thanos.DownsamplingWindow = window
+	meta.Compaction = origMeta.Compaction
+
+	if err := block.WriteMetaFile(bdir, meta); err != nil {
+		return id, errors.Wrap(err, "write block meta")
+	}
+	return id, nil
+}
+
+type aggregator struct {
```

Comment:
It seems you diverge one path for gauge and one for counter. Can we then have separate `aggregator` structs for them?